### PR TITLE
Bitbucket integration

### PR DESCRIPTION
**Pull request by elleobrien**
To do:

- [ ] Logic for detecting CI platform (GitLab, GH, or BB) in `bin/cml-send-comment.js` needs to be decided
- [x] Check tensorboard/cml-publish functions
- [ ] Need a code quality and style check by David!

### CODE REVIEW

File: bin/cml-send-comment.js
Comment (DavidGOrtega) [inline bin/cml-send-comment.js:10]: we have to find a way to determine the env. However this might be useless with the new approach

File: package.json
Comment (DavidGOrtega) [inline package.json:4]: just curiosity, why do you like this format?

File: src/bitbucket.js
Comment (DavidGOrtega) [inline src/bitbucket.js:2]: unused

File: src/bitbucket.js
Comment (DavidGOrtega) [inline src/bitbucket.js:3]: unused

File: src/bitbucket.js
Comment (DavidGOrtega) [inline src/bitbucket.js:8]: unused

File: src/bitbucket.js
Comment (DavidGOrtega) [inline src/bitbucket.js:15]: unused

File: src/bitbucket.js
Comment (DavidGOrtega) [inline src/bitbucket.js:30]: should be using commit_sha, the cml command will set it with BITBUCKET_COMMIT by default since you are exposing BITBUCKET_COMMIT later on as head_sha ;)

File: src/bitbucket.js
Comment (DavidGOrtega) [inline src/bitbucket.js:88]: needs extra line

File: src/bitbucket.js
Comment (DavidGOrtega) [inline src/bitbucket.js:58]: console is intended?

File: src/bitbucket.js
Comment (DavidGOrtega) [inline src/bitbucket.js:67]: console

**Review by DavidGOrtega — COMMENTED** [review]
Looks very good!

Just some comments. Looks like you have issues with eslint, looks like is not working pre-commiting the work. 
There is a tool called husky that is running before the commit and fixing and prettifying the code. Also exposing errors.
Let me show it to you later if you need.

image

**Review by DavidGOrtega — CHANGES_REQUESTED** [review]
Until we fix the import we cant merge this

File: src/drivers/bitbucket_cloud.js
Comment (DavidGOrtega) [inline src/drivers/bitbucket_cloud.js:66]: This makes bitbucket dependant on Env again. If we are in gitlab CI pushing to Bitbucket would need to setup this ENV variable previously.

Comment (DavidGOrtega) [inline src/drivers/bitbucket_cloud.js:66]: This makes bitbucket dependant on Env again. If we are in gitlab CI pushing to Bitbucket would need to setup this ENV variable previously.

File: src/drivers/bitbucket_cloud.js
Comment (DavidGOrtega) [inline src/drivers/bitbucket_cloud.js:69]: Lets require the token full. I mean the user has to provide is the ```repo_token``` or ```BITBUCKET_TOKEN``` as env or as a parameter and thats the base64 of their user and password. They have to calculate manually but I think is fine. What do you think?

```js
const headers = {
      'Content-Type': 'application/json',
      'Authorization': `Basic ${token}`
};
```

I have thought also to extend the input with ```user``` ```pass``` alternatively to ```token``` so we can do it for the user but Im not sure yet. What do you think?

File: package.json
Comment (DavidGOrtega) [inline package.json:60]: This is not needed

**Review by DavidGOrtega — CHANGES_REQUESTED** [review]
Amazing @elleobrien!

Just only one consideration about the token and ENV dependancy.

File: src/drivers/bitbucket_cloud.js
Comment (elleobrien) [reply src/drivers/bitbucket_cloud.js:66]: Hm, if they're in GitLab CI trying to make a request to the BBCloud api, we need some way of getting their BB username. 

We could try using the repo URL, but if they're using an organizational workspace the URL won't necessarily contain their account name. What makes sense?

Assuming we let folks add their username as a secret (when not using BBCloud as driver), I imagine we can do:
```
if (type(repo) == BBCloud){
 const {BITBUCKET_WORKSPACE}=process.env;
 const username = BITBUCKET_WORKSPACE
} else{
 const {BITBUCKET_ACCOUNT_SECRET} = process.env;
 const username = BITBUCKET_ACCOUNT_SECRET
};

```

File: src/drivers/bitbucket_cloud.js
Comment (elleobrien) [reply src/drivers/bitbucket_cloud.js:69]: so you are saying that the user will need to manually generate a token that is 

base64($username:$password)? 

The only concern i have is... will this force them to write a little base64 encoder, or use some online widget to do it? I don't like asking people to manually create the key. I imagine a lot of users getting lost at this step or quitting, unless they'res an easy way to do I don't know. 

Not sure I understand your idea about extending the input, maybe if you make some sample code I could understand better!

File: package.json
Comment (elleobrien) [reply package.json:60]: thanks for catching

File: src/drivers/bitbucket_cloud.js
Comment (DavidGOrtega) [reply src/drivers/bitbucket_cloud.js:69]: @elleobrien 
instead of using token they would have to setup user and password like this

```sh
cml-send-comment --user $MYUSER --pass $MYPASS report.md
```

alternatively they could set 
```sh
cml-send-comment --token $MYTOKEN report.md
```

where token is the base64($username:$password)

File: src/drivers/bitbucket_cloud.js
Comment (DavidGOrtega) [reply src/drivers/bitbucket_cloud.js:66]: > we need some way of getting their BB username.

As input or pre calculated token see below.
We can't depend on variables. The new architecture tries to infer as much as it can but now this driver wont operate in Gitlab CI without setting and ENV var

File: src/drivers/bitbucket_cloud.js
Comment (dmpetrov) [reply src/drivers/bitbucket_cloud.js:69]: I'm trying to understand what is the fundamental issue here. Why BB cannot be implemented the same way as GH/GL with a single token? Will the problem be solved by the separation of BB-cloud & BB-server drivers? I'd appreciate it if you can explain it.

Passwords and secrets in the command line is a potential security breach. In some settings, these passwords will end up in a centralized logging system with wide access across org. It needs to be avoided as much as possible.

**Review by dmpetrov — CHANGES_REQUESTED** [review]
It is an absolutely amazing change ✨ CML is "escaping" GH/GL.

I have a question about token(s) and passwords - see inline

**Comment by elleobrien** [comment]
To answer your question @dmpetrov, BB doesn't provide "personal access
tokens". You need to authenticate with username + password. BUT you can add
those as "masked" secrets, so they're not saved in any logging (i.e., the
BitBucket pipeline logs). So it's no more a security breach than any other
secret.

I think I don't understand what is the problem with using an if-statement.
If the driver is inferred to be BitBucket cloud, we can use an
environmental variable. If not, we can look for user credentials as secrets
(username & password, or one token- base64(username:password). What would
the issue be with that approach?

**Comment by DavidGOrtega** [comment]
@dmpetrov as @elleobrien the user pass or token has to be secrets.
The main issue right now with this PR is that the request is generating the token from an ENV VAR and we dont want to do that anymore. Doing that implies that we have to setup that VAR in another CI ENV or local.

```js
Authorization': 'Basic ' + Base64.encode(`${BITBUCKET_WORKSPACE}:${token}`)};
```

So the solution is expect the token to be generated by them as base64 of user:pass like this:
```
$(echo  $USER_SECRET:$PASS_SECRET | base64)
```

Or introduce the possibility os setting ```user``` and ```password```.
of course @elleobrien we could infer the ```user``` from ```BITBUCKET_WORKSPACE``` but now we are just only reading it from ENV always. If I were using GL as CI I can not comment without setting BITBUCKET_WORKSPACE ENV  VAR

**Comment by elleobrien** [comment]
> @elleobrien of course we could infer the `user` from `BITBUCKET_WORKSPACE` but now we are just only reading it from ENV always. If I were using GL as CI I can not comment without setting BITBUCKET_WORKSPACE ENV VAR.

@DavidGOrtega yes I got this- what I'm wondering is, should we use an if-statement to gate this? So...

if inferred driver != BB:
   look for user ID secret
if inferred driver == BB:
  use environmental variable

The idea being, don't make people add a secret if they don't need to (when BB is the driver). It does look like the GH & GL modules use some environmental variables, but based on what the inferred driver is. 

It sounds like the alternative is that we always make people specify the username & password when they want to leave CML comments on a BB PR, regardless of their driver. Is this what you think is right, @DavidGOrtega? Apologies if I'm not understanding!

**Comment by DavidGOrtega** [comment]
@elleobrien yeah I think that the saver way to do this is just enforce them to setup a token like always with an example.
 
they can generate it with

```
echo  $USER_SECRET:$PASS_SECRET | base64
```

so they just setup token or pass token this way

File: src/cml.js
Comment (DavidGOrtega) [inline src/cml.js:48]: lets add BITBUCKET_TOKEN to maintain the structure

File: src/drivers/bitbucket_cloud.js
Comment (DavidGOrtega) [inline src/drivers/bitbucket_cloud.js:18]: double slash

File: src/drivers/bitbucket_cloud.js
Comment (DavidGOrtega) [inline src/drivers/bitbucket_cloud.js:40]: Does BB support any kind of upload?

**Review by DavidGOrtega — CHANGES_REQUESTED** [review]
Perfect! Just only three silly things

File: src/drivers/bitbucket_cloud.js
Comment (elleobrien) [reply src/drivers/bitbucket_cloud.js:40]: there are artifacts, but they're impermanent- 14 days https://support.atlassian.com/bitbucket-cloud/docs/use-artifacts-in-steps/
Makes me think there are very few advantages to integrating with CML.

File: src/drivers/bitbucket_cloud.js
Comment (elleobrien) [reply src/drivers/bitbucket_cloud.js:18]: is double slash wrong? I assumed it's an escape

Comment (elleobrien) [reply src/drivers/bitbucket_cloud.js:18]: you're right, it's fine to use single slash. but note that i had copied this code from `gitlab.js`- so same double slash is there

File: src/cml.js
Comment (elleobrien) [reply src/cml.js]: done!

File: src/drivers/bitbucket_cloud.js
Comment (DavidGOrtega) [reply src/drivers/bitbucket_cloud.js]: Ah sorry its OK!!!! protocol needs double slash!

**Review by DavidGOrtega — APPROVED** [review]

**Comment by DavidGOrtega** [comment]
Excellent! 🚀

**Pull request merged**
